yarn-project/end-to-end/src/e2e_epochs/epochs_l1_reorgs.test.ts

```diff
@@ -0,0 +1,192 @@
+import { type AztecNode, type Logger, retryUntil } from '@aztec/aztec.js';
+import type { ChainMonitor, Delayer } from '@aztec/ethereum/test';
+import { executeTimeout } from '@aztec/foundation/timer';
+import type { ProverNode } from '@aztec/prover-node';
+
+import { jest } from '@jest/globals';
+import 'jest-extended';
+import { keccak256 } from 'viem';
+
+import type { EndToEndContext } from '../fixtures/utils.js';
+import { EpochsTestContext, L1_BLOCK_TIME_IN_S, L2_SLOT_DURATION_IN_S } from './epochs_test.js';
+
+jest.setTimeout(1000 * 60 * 10);
+
+describe('e2e_epochs/epochs_l1_reorgs', () => {
+  let context: EndToEndContext;
+  let logger: Logger;
+  let node: AztecNode;
+  let proverNode: ProverNode;
+  let monitor: ChainMonitor;
+  let proverDelayer: Delayer;
+  let sequencerDelayer: Delayer;
+
+  let test: EpochsTestContext;
+
+  beforeEach(async () => {
+    test = await EpochsTestContext.setup({
+      l1PublishRetryIntervalMS: 300_000, // Do not retry l1 txs, we dont want them to land
+      txPropagationMaxQueryAttempts: 2, // We are blocking txs here, so do not spend much time looking for them
+    });
+    ({ proverDelayer, sequencerDelayer, context, logger, monitor } = test);
+    node = context.aztecNode;
+    proverNode = context.proverNode!;
+  });
+
+  afterEach(async () => {
+    await test.teardown();
+  });
+
+  it('prunes L2 blocks if a proof is removed due to an L1 reorg', async () => {
+    // Wait until we have proven something and the nodes have caught up
+    logger.warn(`Waiting for initial proof to land`);
+    const provenBlock = await test.waitUntilProvenL2BlockNumber(1);
+    await retryUntil(() => node.getProvenBlockNumber().then(p => p >= provenBlock), 'node sync', 10, 0.1);
+
+    // Stop the prover node so it doesn't re-submit the proof after we've removed it
+    logger.warn(`Proof for block ${provenBlock} mined, stopping prover node`);
+    await proverNode.stop();
+
+    // And remove the proof from L1
+    await context.cheatCodes.eth.reorg(2);
+    await monitor.run();
+    expect(monitor.l2ProvenBlockNumber).toEqual(0);
+
+    // Wait until the end of the proof submission window for the first epoch
+    await test.waitUntilEndOfProofSubmissionWindow(0);
+
+    // Ensure that a new node sees the reorg
+    logger.warn(`Syncing new node to test reorg`);
+    const newNode = await executeTimeout(() => test.createNonValidatorNode(), 10_000, `new node sync`);
+    expect(await newNode.getProvenBlockNumber()).toEqual(0);
+
+    // Latest block number seen by the node may be the current one, or one less if it was *just* mined
```

Comment:
Not fully following here why the span, I am probably just stupid here.

Reply:
Not at all, it took me some time to debug. I've now clarified in the comments:

    // This is because the call to createNonValidatorNode will block until the initial sync is completed,
    // but the initial sync is done to the latest L1 block _at the time the initial sync starts_. So a new
    // node may have appeared while the initial sync runs, that's why we account for a small span of blocks.